### PR TITLE
Lock object optimisation

### PR DESCRIPTION
**Pull request by adamretter**
Backported from FusionDB - improvements in the area of Collection and Document locks - some small reductions in the number of object allocations, less array copies, and ultimately less GC.

### CODE REVIEW

File: exist-core/src/main/java/org/exist/storage/lock/ManagedSingleLockDocumentLock.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    private ManagedSingleLockDocumentLock(final XmldbURI documentUri) {
+        this(documentUri, -1, null, Lock.LockMode.NO_LOCK, null);
+    }
+
+    @Override
+    public void close() {
+        if (!closed) {
+            if (lock != null) {  // NOTE(AR) only null when constructed from {@link #notLocked(XmldbURI)}.
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/storage/lock/ManagedSingleLockDocumentLock.java:63]: 62/63 could be combined? maybe null check is faster than value comparison?

**Comment by dizzzz** [comment]
certain parts of code are much more readable, great.
Are there any new class specific tests missing?

**Review by dizzzz — CHANGES_REQUESTED** [review]
build fails....

File: exist-core/src/main/java/org/exist/storage/lock/ManagedSingleLockDocumentLock.java (hunk 1 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/storage/lock/ManagedSingleLockDocumentLock.java]: Thanks @dizzzz I just pushed a commit to adjust that as you suggested.

**Comment by adamretter** [comment]
> Are there any new class specific tests missing? I guess that these classes are all implicitly tested?

The new classes are just replacements to existing implementations, and so they they are tested by the existing tests.

**Comment by adamretter** [comment]
> build fails....

Should be fixed now I hope.

**Review by reinhapa — APPROVED** [review]

**Review by dizzzz — APPROVED** [review]

**Comment by adamretter** [comment]
Thanks @dizzzz

**Pull request merged**